Continue support sharding pipes in `tud.datapipes.iter.grouping` as deprecated

wenleix opened the pull request:
Summary:
https://github.com/pytorch/pytorch/pull/94095 moves this into `tud.datapipes.iter.sharding`. However, since previously this is a public API, this is a BC break change.

As discussed in https://github.com/pytorch/data/pull/987#issuecomment-1422440049, we will have backward compatbile support but with deprecated warning.

Differential Revision: D43161015

---
Deprecation:
Please import from `SHARDING_PRIORITIES` from `torch.utils.data.datapipes.iter.sharding` instead of `torch.utils.data.datapipes.iter.grouping`.

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -11,7 +14,16 @@
+
+def __getattr__(name: str):
+    if name in ["SHARDING_PRIORITIES", "ShardingFilterIterDataPipe"]:
+        warnings.warn(f"Please use `{name}` from the `torch.utils.data.datapipes.iter.sharding` package",
+                      category=FutureWarning, stacklevel=2)
```
Can we add one more comment saying it's going to be removed in PyTorch 2.1?

NivekT replied:
nit: I notice that TorchData imports from `torch.utils.data.datapipes.iter` for `ShardingFilter` and `torch.utils.data.datapipes.iter.sharding` for `SHARDING_PRIORITIES`.

Should we have separate comments for each `name`?

wenleix replied:
> I notice that TorchData imports from torch.utils.data.datapipes.iter for ShardingFilter and torch.utils.data.datapipes.iter.sharding for SHARDING_PRIORITIES.

Good point... I think `tudd.iter.ShardingFilter` is preferred over `tudd.iter.sharding.ShardingFilterIterDataPipe`.

But if user already used `tudd.iter.grouping.ShardingFilterIterDataPipe`, perhaps the immediate rerouting would be `tudd.iter.sharding.ShardingFilterIterDataPipe` ~

NivekT replied:
I think we should do two warnings here because users probably won't be happy to see another warning after changing the code to the proposed fix.

wenleix replied:
Do we have a guideline on this? Since `tudd.iter.sharding.ShardingFilterIterDataPipe` is also a public API. While we encourage/promote to use `tudd.iter.ShardingFilter`, both should be supported.

If we don't want to support `tudd.iter.sharding.ShardingFilterIterDataPipe` (and should always use the `tud.iter` version, shall make it explicitly make it like  `tudd.iter.sharding._ShardingFilterIterDataPipe`? (Similar question applies to other DataPipes under `tudd.iter.XXX`?

ejguan commented on test/test_datapipe.py:
```diff
@@ -2722,6 +2722,45 @@ def construct_sharded_pipe():
+    # Test tud.datapipes.iter.grouping.SHARDING_PRIORITIES for backward compatbility
+    # TODO: Remove this test once tud.datapipes.iter.grouping.SHARDING_PRIORITIES is deprecated
+    def test_sharding_groups_in_legacy_grouping_package(self):
+        from torch.utils.data.datapipes.iter.grouping import SHARDING_PRIORITIES as LEGACY_SHARDING_PRIORITIES
```
Should we add a `assertWarning` at import?
And, other tests can be removed as they just references to the right classes.

wenleix replied:
> And, other tests can be removed as they just references to the right classes.

discussed offline, just use them as some smoke integration test~

NivekT reviewed (commented):
Didn't realize `__getattr__` worked for modules. LGTM after resolving the comments!

wenleix commented:
> Didn't realize __getattr__ worked for modules. LGTM after resolving the comments!

PEP-562: https://peps.python.org/pep-0562/ ;)

wenleix commented:
@pytorchbot merge

wenleix committed:
Continue support sharding pipes in `tud.datapipes.iter.grouping` as deprecated (#94527)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/94527

https://github.com/pytorch/pytorch/pull/94095 moves this into `tud.datapipes.iter.sharding`. However, since previously this is a public API, this is a BC break change.

As discussed in https://github.com/pytorch/data/pull/987#issuecomment-1422440049, we will have backward compatbile support but with deprecated warning.

Test Plan:
```
buck test mode/opt //caffe2/test:datapipe
```

Reviewed By: ejguan

Differential Revision: D43161015

fbshipit-source-id: 82049979981b83aefd247843fbc2c004a7e7a90a
